### PR TITLE
Revert PR 12684 and 12556

### PR DESCRIPTION
**Pull request by hx235**
**Context/Summary:** a better API design is decided lately so we decided to revert these two changes.

**Test plan:**
- CI

### CODE REVIEW

File: file/writable_file_writer.h [hunk 1]
```diff
@@ -297,7 +304,9 @@ class WritableFileWriter {
   }
   void set_seen_error() { seen_error_.store(true, std::memory_order_relaxed); }
 
-  IOStatus GetWriterHasPreviousErrorStatus() {
+  IOStatus AssertFalseAndGetStatusForPrevError() {
+    // This should only happen if SyncWithoutFlush() was called.
+    assert(sync_without_flush_called_);
```
Comment (ajkr) [inline file/writable_file_writer.h:309]: This adds back an arbitrary restriction in a set of circumstances (no `SyncWithoutFlush()`) that feels totally unrelated to the restriction itself. It took a while to remove it despite all that. Can we leave it deleted?

**Review by ajkr — CHANGES_REQUESTED** [review]

File: file/writable_file_writer.h (hunk 1 above)
Comment (hx235) [reply file/writable_file_writer.h:309]: Sure - sorry I missed this part of the PR

Comment (hx235) [reply file/writable_file_writer.h]: Fixed

**Comment by hx235** [comment]
Rebasing

File: db/db_impl/db_impl.cc [hunk 2]
```diff
@@ -650,7 +637,7 @@ Status DBImpl::CloseHelper() {
       if (!s.ok()) {
         ROCKS_LOG_WARN(
             immutable_db_options_.info_log,
-            "Unable to clear writer for WAL %s with error -- %s",
+            "Unable to Sync WAL file %s with error -- %s",
```
Comment (ajkr) [inline db/db_impl/db_impl.cc:640]: Is this correct? It looks like this is the status for `log.ClearWriter()` (I didn't look at what it does).

**Review by ajkr — APPROVED** [review]
LGTM

File: db/db_impl/db_impl.cc (hunk 2 above)
Comment (hx235) [reply db/db_impl/db_impl.cc:640]: It was the original message - most likely incorrect - will fix it.

Comment (hx235) [reply db/db_impl/db_impl.cc]: Fixed
